src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodFilteringService.cs

```diff
@@ -0,0 +1,191 @@
+        {
+            var client = await TryGetRemoteHostClientAsync(project, cancellationToken).ConfigureAwait(false);
+            if (client == null)
+            {
+                return await GetPossibleExtensionMethodMatchesInCurrentProcessAsync(
+                    project, targetTypeNames, loadOnly, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                return await GetPossibleExtensionMethodMatchesInRemoteProcessAsync(
+                    client, project, targetTypeNames, loadOnly, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        private static async Task<MultiDictionary<string, string>?> GetPossibleExtensionMethodMatchesInRemoteProcessAsync(RemoteHostClient client, Project project, ImmutableHashSet<string> targetTypeNames, bool loadOnly, CancellationToken cancellationToken)
+        {
+            var remoteResults = await client.TryRunCodeAnalysisRemoteAsync<IEnumerable<(string, IEnumerable<string>)>>(
+                project.Solution,
+                nameof(IRemoteExtensionMethodFilteringService.GetPossibleExtensionMethodMatchesAsync),
+                new object[] { project.Id, targetTypeNames.ToArray(), loadOnly },
+                cancellationToken).ConfigureAwait(false);
+
+            if (!remoteResults.Any())
+            {
+                return null;
+            }
+
+            // Reconstruct the dictionary from remote results.
+            var results = new MultiDictionary<string, string>();
+
+            foreach (var pair in remoteResults)
+            {
+                foreach (var value in pair.Item2)
+                {
+                    results.Add(pair.Item1, value);
+                }
+            }
+
+            return results;
+        }
+
+        private static async Task<MultiDictionary<string, string>?> GetPossibleExtensionMethodMatchesInCurrentProcessAsync(Project currentProject, ImmutableHashSet<string> targetTypeNames, bool loadOnly, CancellationToken cancellationToken)
+        {
+            var solution = currentProject.Solution;
+            var graph = currentProject.Solution.GetProjectDependencyGraph();
+            var relevantProjectIds = graph.GetProjectsThatThisProjectTransitivelyDependsOn(currentProject.Id)
+                                        .Concat(currentProject.Id);
+
+            using var syntaxDisposer = ArrayBuilder<SyntaxTreeIndex>.GetInstance(out var syntaxTreeIndexBuilder);
+            using var symbolDisposer = ArrayBuilder<SymbolTreeInfo>.GetInstance(out var symbolTreeInfoBuilder);
+            var peReferences = PooledHashSet<PortableExecutableReference>.GetInstance();
+
+            try
+            {
+                foreach (var projectId in relevantProjectIds.Concat(currentProject.Id))
+                {
+                    // Alway create indices for documents in current project if they don't exist.
+                    loadOnly &= projectId != currentProject.Id;
```

Comment:
assigning over loadOnly seems bad.  it seems like it wouldn't actually work since you'll set it to false, then get to currentProject, and try to & in 'true'.

Reply:
Fixed